apps/api/src/http/internal/rest/webhooks/stripe/subscription.rs

```diff
@@ -0,0 +1,253 @@
+use std::collections::HashMap;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use mongodb::options::{FindOneAndUpdateOptions, ReturnDocument};
+use shared::database::product::subscription::{
+	ProviderSubscriptionId, SubscriptionId, SubscriptionPeriod, SubscriptionPeriodCreatedBy,
+};
+use shared::database::product::{ProductId, SubscriptionProduct, SubscriptionProductVariant};
+use shared::database::queries::{filter, update};
+use shared::database::user::UserId;
+
+use crate::global::Global;
+use crate::http::error::ApiError;
+use crate::transactions::{TransactionError, TransactionResult, TransactionSession};
+
+fn subscription_products(items: stripe::List<stripe::SubscriptionItem>) -> Result<Vec<ProductId>, ApiError> {
+	items
+		.data
+		.into_iter()
+		.map(|i| Ok(ProductId::from(i.price.ok_or(ApiError::BAD_REQUEST)?.id)))
+		.collect()
+}
+
+/// Creates the subscription object in the database.
+pub async fn created(
+	_global: &Arc<Global>,
+	mut tx: TransactionSession<'_, ApiError>,
+	subscription: stripe::Subscription,
+) -> TransactionResult<Option<SubscriptionId>, ApiError> {
+	let items = subscription_products(subscription.items).map_err(TransactionError::custom)?;
+
+	let products = tx
+		.find(
+			filter::filter! {
+				SubscriptionProduct {
+					#[query(flatten)]
+					variants: SubscriptionProductVariant {
+						#[query(selector = "in", serde)]
+						id: items,
+					}
+				}
+			},
+			None,
+		)
+		.await?;
+
+	if products.len() != 1 {
+		// only accept subs with one product
+		return Ok(None);
+	}
```

Comment:
Perhaps we should log if there are multiple matches? Also what happens if more than one sub product matches the same id? do we outlaw that in a constraint?